src/coord/src/coord.rs

```diff
@@ -817,29 +846,67 @@ impl Coordinator {
                             );
                         }
                         if index_state.since != compaction_frontier {
+                            // Mark previous since for eviction.
+                            let mut updates: Vec<_> = index_state
+                                .since
+                                .elements()
+                                .iter()
+                                .map(|time| (*time, -1))
+                                .collect();
                             index_state.advance_since(&compaction_frontier);
-                            self.since_updates
-                                .push((name.clone(), index_state.since.clone()));
+                            // Mark new since for insertion.
+                            for time in index_state.since.elements() {
+                                updates.push((*time, 1));
+                            }
+                            // We always want `compaction_frontiers` to have exactly one timestamp tracking
+                            // the index's since, so send both the eviction and insertion in the same
+                            // update batch.
+                            self.update_compaction_frontier(*name, updates);
                         }
                     }
                 }
             }
         }
     }
 
+    /// Updates the tracked compaction frontier for an index.
+    ///
+    /// If the antichain progressed, it is added to the pending compactions.
+    fn update_compaction_frontier<I>(&mut self, name: GlobalId, updates: I)
+    where
+        I: IntoIterator<Item = (Timestamp, i64)>,
+    {
+        // Indexes are the only things that can be compacted.
+        assert!(self.catalog.get_by_id(&name).item_type() == CatalogItemType::Index);
+        // Indexes added via `Frontiers::new` get their `since` set to
+        // `Timestamp::minimum()`. Mirror that here when we encounter an unknown index.
+        let entry = self
+            .compaction_frontiers
+            .entry(name)
+            .or_insert_with(|| MutableAntichain::new_bottom(Timestamp::minimum()));
+        let before = entry.frontier().to_owned();
+        let changed = entry.update_iter(updates).next().is_some();
+        let after = entry.frontier().to_owned();
+        // If the antichain retreated, we've done something wrong.
+        assert!(
+            timely::order::PartialOrder::less_equal(&before, &after),
+            "{}: {:?} > {:?}",
```

Comment:
Minor, but more idiomatic timely to report `!<=`, however you'd like to type it, as we only know that the `less_equal` test failed not that the opposite order holds.